File: frameos/src/drivers/waveshare/ePaper/EPD_4in0e.nim
## ***************************************************************************
##  | File        :   EPD_4in0e.nim
##  | Author      :   Waveshare team (original C implementation)
##  | Nim Port    :   FrameOS maintainers
##  | Function    :   4inch e-Paper (E) Driver
##  | Info        :   Native Nim implementation of the Waveshare driver
## ----------------
##  | This version:   V1.0
##  | Date        :   2024-08-20
##  | Info        :
##  -----------------------------------------------------------------------------
## #
## # Permission is hereby granted, free of charge, to any person obtaining a copy
## # of this software and associated documnetation files (the "Software"), to deal
## # in the Software without restriction, including without limitation the rights
## # to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
## # copies of the Software, and to permit persons to  whom the Software is
## # furished to do so, subject to the following conditions:
## #
## # The above copyright notice and this permission notice shall be included in
## # all copies or substantial portions of the Software.
## #
## # THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
## # IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
## # FITNESS OR FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
## # AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
## # LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
## # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
## # THE SOFTWARE.
## #
## ****************************************************************************

import
  DEV_Config,
  json,
  options,
  strformat,
  strutils,
  times,
  std/monotimes

from drivers/waveshare/types import logDriverDebug, driverDebugLogsEnabled
import frameos/utils/time

const
  EPD_4IN0E_WIDTH* = 400
  EPD_4IN0E_HEIGHT* = 600
  busyLogLoopInterval = 100
  busyLogMinIntervalMs = 1000.0

## ********************************
## Color Index
## ********************************

const
  EPD_4IN0E_BLACK* = 0x0
  EPD_4IN0E_WHITE* = 0x1
  EPD_4IN0E_YELLOW* = 0x2
  EPD_4IN0E_RED* = 0x3
  EPD_4IN0E_BLUE* = 0x5
  EPD_4IN0E_GREEN* = 0x6

type
  UByteArray = UncheckedArray[UBYTE]

var
  dataLogCounter = 0
  dataBytesCurrentCommand = 0

proc capturePinStates(): JsonNode =
  %*{
    "busy": DEV_Digital_Read(UWORD(EPD_BUSY_PIN)).int,
    "rst": DEV_Digital_Read(UWORD(EPD_RST_PIN)).int,
    "dc": DEV_Digital_Read(UWORD(EPD_DC_PIN)).int,
    "cs": DEV_Digital_Read(UWORD(EPD_CS_PIN)).int,
    "pwr": DEV_Digital_Read(UWORD(EPD_PWR_PIN)).int
  }

template logDebug(action: string, extra: JsonNode = nil) =
  if driverDebugLogsEnabled():
    var payload = %*{"event": "driver:waveshare:debug", "action": action}
    if extra != nil and extra.kind == JObject:
      for key, value in extra.pairs:
        payload[key] = value
    logDriverDebug(payload)

proc logCommand(reg: UBYTE) =
  dataLogCounter = 0
  dataBytesCurrentCommand = 0
  if driverDebugLogsEnabled():
    logDriverDebug(%*{
      "event": "driver:waveshare:command",
      "command": reg.int,
      "commandHex": &"0x{toHex(reg.int, 2)}"
    })

proc epd4in0eReset() =
  logDebug("reset:start")
  DEV_Digital_Write(UWORD(EPD_RST_PIN), UBYTE(1))
  DEV_Delay_ms(UDOUBLE(20))
  DEV_Digital_Write(UWORD(EPD_RST_PIN), UBYTE(0))
  DEV_Delay_ms(UDOUBLE(2))
  DEV_Digital_Write(UWORD(EPD_RST_PIN), UBYTE(1))
  DEV_Delay_ms(UDOUBLE(20))
  logDebug("reset:done")

proc epd4in0eSendCommand(reg: UBYTE) =
  logCommand(reg)
  DEV_Digital_Write(UWORD(EPD_DC_PIN), UBYTE(0))
  DEV_Digital_Write(UWORD(EPD_CS_PIN), UBYTE(0))
  DEV_SPI_WriteByte(reg)
  DEV_Digital_Write(UWORD(EPD_CS_PIN), UBYTE(1))

proc epd4in0eSendData(data: UBYTE) =
  DEV_Digital_Write(UWORD(EPD_DC_PIN), UBYTE(1))
  DEV_Digital_Write(UWORD(EPD_CS_PIN), UBYTE(0))
  DEV_SPI_WriteByte(data)
  DEV_Digital_Write(UWORD(EPD_CS_PIN), UBYTE(1))

  inc dataBytesCurrentCommand
  if driverDebugLogsEnabled():
    if dataLogCounter < 16:
      logDriverDebug(%*{
        "event": "driver:waveshare:data",
        "index": dataBytesCurrentCommand,
        "data": data.int,
        "dataHex": &"0x{toHex(data.int, 2)}"
      })
    elif dataLogCounter == 16:
      logDriverDebug(%*{
        "event": "driver:waveshare:data",
        "message": "Further data logging suppressed for this command",
        "bytesSent": dataBytesCurrentCommand
      })
    elif (dataBytesCurrentCommand mod 4096) == 0:
      logDriverDebug(%*{
        "event": "driver:waveshare:data",
        "message": "Data transfer progress",
        "bytesSent": dataBytesCurrentCommand
      })
    inc dataLogCounter

proc epd4in0eReadBusyH() =
  let startTime = getMonoTime()
  var loopCount = 0
  var lastLog = startTime
  let initialState = DEV_Digital_Read(UWORD(EPD_BUSY_PIN))
  logDebug("busy:wait:start", %*{
    "initialState": initialState.int,
    "pins": capturePinStates()
  })

  var observedLow = initialState == UBYTE(0)
  var lowStartTime = if observedLow: some(startTime) else: none(MonoTime)

  if (not observedLow) and DEV_Digital_Read(UWORD(EPD_BUSY_PIN)) == UBYTE(0):
    observedLow = true
    lowStartTime = some(getMonoTime())

  while DEV_Digital_Read(UWORD(EPD_BUSY_PIN)) == UBYTE(0):
    if not observedLow:
      observedLow = true
      lowStartTime = some(getMonoTime())

    DEV_Delay_ms(UDOUBLE(10))
    inc loopCount

    if driverDebugLogsEnabled() and loopCount mod busyLogLoopInterval == 0:
      let now = getMonoTime()
      if durationToMilliseconds(now - lastLog) >= busyLogMinIntervalMs:
        logDriverDebug(%*{
          "event": "driver:waveshare:busy",
          "loops": loopCount,
          "elapsedMs": durationToMilliseconds(now - startTime),
          "stage": "waitForHigh",
          "pins": capturePinStates()
        })
        lastLog = now

  DEV_Delay_ms(UDOUBLE(100))

  let endTime = getMonoTime()
  let durationMs = durationToMilliseconds(endTime - startTime)
  let finalState = DEV_Digital_Read(UWORD(EPD_BUSY_PIN))
  let finalPins = capturePinStates()

  let waitForLowMs =
    if observedLow and lowStartTime.isSome:
      durationToMilliseconds(lowStartTime.get() - startTime)
    else: 0.0
  let waitForHighMs =
    if observedLow and lowStartTime.isSome:
      durationToMilliseconds(endTime - lowStartTime.get())
    else: 0.0

  logDebug("busy:wait:end", %*{
    "durationMs": durationMs,
    "loops": loopCount,
    "finalState": finalState.int,
    "observedLow": observedLow,
    "waitedForLowMs": waitForLowMs,
    "waitedForHighMs": waitForHighMs,
    "timedOutWaitingForLow": false,
    "timedOutWaitingForHigh": false,
    "pins": finalPins
  })

proc epd4in0eTurnOnDisplay() =
  logDebug("turnOnDisplay:start")
  logDebug("turnOnDisplay:powerOn")
  epd4in0eSendCommand(0x04) # POWER_ON
  epd4in0eReadBusyH()
  DEV_Delay_ms(UDOUBLE(200))

  logDebug("turnOnDisplay:secondSetting")
  ## Second setting
  epd4in0eSendCommand(0x06)
  epd4in0eSendData(0x6F)
  epd4in0eSendData(0x1F)
  epd4in0eSendData(0x17)
  epd4in0eSendData(0x27)
  DEV_Delay_ms(UDOUBLE(200))

  logDebug("turnOnDisplay:refresh")
  epd4in0eSendCommand(0x12) # DISPLAY_REFRESH
  epd4in0eSendData(0x00)
  epd4in0eReadBusyH()

  logDebug("turnOnDisplay:powerOff")
  epd4in0eSendCommand(0x02) # POWER_OFF
  epd4in0eSendData(0x00)
  epd4in0eReadBusyH()
  logDebug("turnOnDisplay:done")

proc EPD_4IN0E_Init*() =
  logDebug("init:start")
  epd4in0eReset()
  epd4in0eReadBusyH()
  DEV_Delay_ms(UDOUBLE(30))
  logDebug("init:afterResetDelay")

  logDebug("init:cmdh")
  epd4in0eSendCommand(0xAA)
  epd4in0eSendData(0x49)
  epd4in0eSendData(0x55)
  epd4in0eSendData(0x20)
  epd4in0eSendData(0x08)
  epd4in0eSendData(0x09)
  epd4in0eSendData(0x18)

  logDebug("init:drvPLL")
  epd4in0eSendCommand(0x01)
  epd4in0eSendData(0x3F)

  logDebug("init:powerSetting")
  epd4in0eSendCommand(0x00)
  epd4in0eSendData(0x5F)
  epd4in0eSendData(0x69)

  logDebug("init:powerOptimisation1")
  epd4in0eSendCommand(0x05)
  epd4in0eSendData(0x40)
  epd4in0eSendData(0x1F)
  epd4in0eSendData(0x1F)
  epd4in0eSendData(0x2C)

  logDebug("init:powerOptimisation2")
  epd4in0eSendCommand(0x08)
  epd4in0eSendData(0x6F)
  epd4in0eSendData(0x1F)
  epd4in0eSendData(0x1F)
  epd4in0eSendData(0x22)

  logDebug("init:powerOptimisation3")
  epd4in0eSendCommand(0x06)
  epd4in0eSendData(0x6F)
  epd4in0eSendData(0x1F)
  epd4in0eSendData(0x17)
  epd4in0eSendData(0x17)

  logDebug("init:boosterSoftStart")
  epd4in0eSendCommand(0x03)
  epd4in0eSendData(0x00)
  epd4in0eSendData(0x54)
  epd4in0eSendData(0x00)
  epd4in0eSendData(0x44)

  logDebug("init:tconSetting")
  epd4in0eSendCommand(0x60)
  epd4in0eSendData(0x02)
  epd4in0eSendData(0x00)

  logDebug("init:powerOptimisation4")
  epd4in0eSendCommand(0x30)
  epd4in0eSendData(0x08)

  logDebug("init:vcomAndDataInterval")
  epd4in0eSendCommand(0x50)
  epd4in0eSendData(0x3F)

  logDebug("init:resolution")
  epd4in0eSendCommand(0x61)
  epd4in0eSendData(0x01)
  epd4in0eSendData(0x90)
  epd4in0eSendData(0x02)
  epd4in0eSendData(0x58)

  logDebug("init:pllControl")
  epd4in0eSendCommand(0xE3)
  epd4in0eSendData(0x2F)

  logDebug("init:vdcsSetting")
  epd4in0eSendCommand(0x84)
  epd4in0eSendData(0x01)
  epd4in0eReadBusyH()
  logDebug("init:done")

proc EPD_4IN0E_Clear*(color: UBYTE) =
  let width =
    if (EPD_4IN0E_WIDTH and 1) == 0:
      EPD_4IN0E_WIDTH div 2
    else:
      EPD_4IN0E_WIDTH div 2 + 1
  let height = EPD_4IN0E_HEIGHT
  let totalBytes = width * height

  logDebug("clear:start", %*{"color": color.int, "widthBytes": width, "height": height, "totalBytes": totalBytes})

  epd4in0eSendCommand(0x10)
  for _ in 0 ..< height:
    for _ in 0 ..< width:
      let packed = (color shl 4) or color
      epd4in0eSendData(UBYTE(packed))

  logDebug("clear:dataWritten", %*{"totalBytes": totalBytes})
  epd4in0eTurnOnDisplay()

proc EPD_4IN0E_Show7Block*() =
  const colorSeven = [
    UBYTE(EPD_4IN0E_BLACK),
    UBYTE(EPD_4IN0E_YELLOW),
    UBYTE(EPD_4IN0E_RED),
    UBYTE(EPD_4IN0E_BLUE),
    UBYTE(EPD_4IN0E_GREEN),
    UBYTE(EPD_4IN0E_WHITE)
  ]

  logDebug("show7Block:start", %*{"blocks": colorSeven.len, "bytesPerBlock": 20000})
  epd4in0eSendCommand(0x10)
  for color in colorSeven:
    for _ in 0 ..< 20000:
      let packed = (color shl 4) or color
      epd4in0eSendData(UBYTE(packed))

  logDebug("show7Block:dataWritten", %*{"totalBytes": colorSeven.len * 20000})
  epd4in0eTurnOnDisplay()

proc EPD_4IN0E_Show*() =
  const colorSeven = [
    UBYTE(EPD_4IN0E_BLACK),
    UBYTE(EPD_4IN0E_YELLOW),
    UBYTE(EPD_4IN0E_RED),
    UBYTE(EPD_4IN0E_BLUE),
    UBYTE(EPD_4IN0E_GREEN),
    UBYTE(EPD_4IN0E_WHITE)
  ]

  let width =
    if (EPD_4IN0E_WIDTH and 1) == 0:
      EPD_4IN0E_WIDTH div 2
    else:
      EPD_4IN0E_WIDTH div 2 + 1
  let height = EPD_4IN0E_HEIGHT
  let totalBytes = width * height

  var k = 0
  var o = 0

  logDebug("show:start", %*{"widthBytes": width, "height": height, "totalBytes": totalBytes})
  epd4in0eSendCommand(0x10)
  for j in 0 ..< height:
    if (j > 10) and (j < 50):
      for _ in 0 ..< width:
        let color = colorSeven[0]
        let packed = (color shl 4) or color
        epd4in0eSendData(UBYTE(packed))
    elif o < height div 2:
      for _ in 0 ..< width:
        let color = colorSeven[0]
        let packed = (color shl 4) or color
        epd4in0eSendData(UBYTE(packed))
    else:
      for _ in 0 ..< width:
        let color = colorSeven[k]
        let packed = (color shl 4) or color
        epd4in0eSendData(UBYTE(packed))
      inc k
      if k >= colorSeven.len:
        k = 0

    inc o
    if o >= height:
      o = 0

  logDebug("show:dataWritten", %*{"totalBytes": totalBytes})
  epd4in0eTurnOnDisplay()

proc EPD_4IN0E_Display*(image: ptr UBYTE) =
  if image.isNil:
    logDebug("display:image:nil")
    return

  let width =
    if (EPD_4IN0E_WIDTH and 1) == 0:
      EPD_4IN0E_WIDTH div 2
    else:
      EPD_4IN0E_WIDTH div 2 + 1
  let height = EPD_4IN0E_HEIGHT
  let buffer = cast[ptr UByteArray](image)
  let totalBytes = width * height

  logDebug("display:start", %*{"widthBytes": width, "height": height, "totalBytes": totalBytes})

  if driverDebugLogsEnabled() and totalBytes > 0:
    let previewCount = min(totalBytes, 16)
    var preview = newSeq[int](previewCount)
    for i in 0 ..< previewCount:
      preview[i] = buffer[i].int
    logDriverDebug(%*{
      "event": "driver:waveshare:dataPreview", "count": previewCount,
      "bytes": preview
    })

  epd4in0eSendCommand(0x10)
  for j in 0 ..< height:
    for i in 0 ..< width:
      let idx = i + j * width
      epd4in0eSendData(buffer[idx])

  logDebug("display:dataWritten", %*{"totalBytes": totalBytes})
  epd4in0eTurnOnDisplay()

proc EPD_4IN0E_Sleep*() =
  logDebug("sleep:start")
  epd4in0eSendCommand(0x07) # DEEP_SLEEP
  epd4in0eSendData(0xA5)
  logDebug("sleep:done")

File: frameos/src/drivers/waveshare/ePaper/EPD_4in0e.c
/*****************************************************************************
* | File        :   EPD_4in0e.c
* | Author      :   Waveshare team
* | Function    :   4inch e-Paper (E) Driver
* | Info        :
*----------------
* | This version:   V1.0
* | Date        :   2024-08-20
* | Info        :
* -----------------------------------------------------------------------------
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
******************************************************************************/
#include "EPD_4in0e.h"
#include "Debug.h"

/******************************************************************************
function :  Software reset
parameter:
******************************************************************************/
static void EPD_4IN0E_Reset(void)
{
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(20);
    DEV_Digital_Write(EPD_RST_PIN, 0);
    DEV_Delay_ms(2);
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(20);
}

/******************************************************************************
function :  send command
parameter:
     Reg : Command register
******************************************************************************/
static void EPD_4IN0E_SendCommand(UBYTE Reg)
{
    DEV_Debug_Command(Reg);
    DEV_Digital_Write(EPD_DC_PIN, 0);
    DEV_Digital_Write(EPD_CS_PIN, 0);
    DEV_SPI_WriteByte(Reg);
    DEV_Digital_Write(EPD_CS_PIN, 1);
}

/******************************************************************************
function :  send data
parameter:
    Data : Write data
******************************************************************************/
static void EPD_4IN0E_SendData(UBYTE Data)
{
    DEV_Debug_Data(Data);
    DEV_Digital_Write(EPD_DC_PIN, 1);
    DEV_Digital_Write(EPD_CS_PIN, 0);
    DEV_SPI_WriteByte(Data);
    DEV_Digital_Write(EPD_CS_PIN, 1);
}

/******************************************************************************
function :  Wait until the busy_pin goes LOW
parameter:
******************************************************************************/
static void EPD_4IN0E_ReadBusyH(const char *stage)
{
    Debug("e-Paper busy H\r\n");
    /* FrameOS divergence: bounded, narrated wait (DEV_Busy_Wait). */
    if (DEV_Busy_Wait(stage, 0, 10) < 0) {
        return;
    }
    DEV_Delay_ms(100);
    Debug("e-Paper busy H release\r\n");
}

/******************************************************************************
function :  Turn On Display
parameter:
******************************************************************************/
static void EPD_4IN0E_TurnOnDisplay(void)
{
    
    DEV_Debug_Log("turnOnDisplay:powerOn", NULL);
    EPD_4IN0E_SendCommand(0x04); // POWER_ON
    EPD_4IN0E_ReadBusyH("turnOnDisplay:powerOn");
    DEV_Delay_ms(200);

    //Second setting 
    DEV_Debug_Log("turnOnDisplay:secondSetting", NULL);
    EPD_4IN0E_SendCommand(0x06);
    EPD_4IN0E_SendData(0x6F);
    EPD_4IN0E_SendData(0x1F);
    EPD_4IN0E_SendData(0x17);
    EPD_4IN0E_SendData(0x27);
    DEV_Delay_ms(200);

    DEV_Debug_Log("turnOnDisplay:refresh", NULL);
    EPD_4IN0E_SendCommand(0x12); // DISPLAY_REFRESH
    EPD_4IN0E_SendData(0x00);
    EPD_4IN0E_ReadBusyH("turnOnDisplay:refresh");

    DEV_Debug_Log("turnOnDisplay:powerOff", NULL);
    EPD_4IN0E_SendCommand(0x02); // POWER_OFF
    EPD_4IN0E_SendData(0X00);
    EPD_4IN0E_ReadBusyH("turnOnDisplay:powerOff");
    DEV_Debug_Log("turnOnDisplay:done", NULL);
}

/******************************************************************************
function :  Initialize the e-Paper register
parameter:
******************************************************************************/
void EPD_4IN0E_Init(void)
{
    DEV_Debug_Log("init:start", NULL);
    EPD_4IN0E_Reset();
    EPD_4IN0E_ReadBusyH("init:reset");
    DEV_Delay_ms(30);

    DEV_Debug_Log("init:cmdh", NULL);
    EPD_4IN0E_SendCommand(0xAA);    // CMDH
    EPD_4IN0E_SendData(0x49);
    EPD_4IN0E_SendData(0x55);
    EPD_4IN0E_SendData(0x20);
    EPD_4IN0E_SendData(0x08);
    EPD_4IN0E_SendData(0x09);
    EPD_4IN0E_SendData(0x18);

    DEV_Debug_Log("init:drvPLL", NULL);
    EPD_4IN0E_SendCommand(0x01);
    EPD_4IN0E_SendData(0x3F);

    DEV_Debug_Log("init:powerSetting", NULL);
    EPD_4IN0E_SendCommand(0x00);
    EPD_4IN0E_SendData(0x5F);
    EPD_4IN0E_SendData(0x69);

    DEV_Debug_Log("init:powerOptimisation1", NULL);
    EPD_4IN0E_SendCommand(0x05);
    EPD_4IN0E_SendData(0x40);
    EPD_4IN0E_SendData(0x1F);
    EPD_4IN0E_SendData(0x1F);
    EPD_4IN0E_SendData(0x2C);

    DEV_Debug_Log("init:powerOptimisation3", NULL);
    EPD_4IN0E_SendCommand(0x08);
    EPD_4IN0E_SendData(0x6F);
    EPD_4IN0E_SendData(0x1F);
    EPD_4IN0E_SendData(0x1F);
    EPD_4IN0E_SendData(0x22);

    DEV_Debug_Log("init:powerOptimisation2", NULL);
    EPD_4IN0E_SendCommand(0x06);
    EPD_4IN0E_SendData(0x6F);
    EPD_4IN0E_SendData(0x1F);
    EPD_4IN0E_SendData(0x17);
    EPD_4IN0E_SendData(0x17);

    DEV_Debug_Log("init:boosterSoftStart", NULL);
    EPD_4IN0E_SendCommand(0x03);
    EPD_4IN0E_SendData(0x00);
    EPD_4IN0E_SendData(0x54);
    EPD_4IN0E_SendData(0x00);
    EPD_4IN0E_SendData(0x44); 

    DEV_Debug_Log("init:resolution", NULL);
    EPD_4IN0E_SendCommand(0x60);
    EPD_4IN0E_SendData(0x02);
    EPD_4IN0E_SendData(0x00);

    DEV_Debug_Log("init:powerOptimisation4", NULL);
    EPD_4IN0E_SendCommand(0x30);
    EPD_4IN0E_SendData(0x08);

    DEV_Debug_Log("init:vcomAndDataInterval", NULL);
    EPD_4IN0E_SendCommand(0x50);
    EPD_4IN0E_SendData(0x3F);

    DEV_Debug_Log("init:resolutionSetting", NULL);
    EPD_4IN0E_SendCommand(0x61);
    EPD_4IN0E_SendData(0x01);
    EPD_4IN0E_SendData(0x90);
    EPD_4IN0E_SendData(0x02); 
    EPD_4IN0E_SendData(0x58);

    DEV_Debug_Log("init:pllControl", NULL);
    EPD_4IN0E_SendCommand(0xE3);
    EPD_4IN0E_SendData(0x2F);

    DEV_Debug_Log("init:vdcsSetting", NULL);
    EPD_4IN0E_SendCommand(0x84);
    EPD_4IN0E_SendData(0x01);
    EPD_4IN0E_ReadBusyH("init:done");
    DEV_Debug_Log("init:done", NULL);
}

/******************************************************************************
function :  Clear screen
parameter:
******************************************************************************/
void EPD_4IN0E_Clear(UBYTE color)
{
    UWORD Width, Height;
    Width = (EPD_4IN0E_WIDTH % 2 == 0)? (EPD_4IN0E_WIDTH / 2 ): (EPD_4IN0E_WIDTH / 2 + 1);
    Height = EPD_4IN0E_HEIGHT;

    DEV_Debug_Log("clear:start", NULL);
    EPD_4IN0E_SendCommand(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_4IN0E_SendData((color<<4)|color);
        }
    }

    DEV_Debug_Log("clear:dataWritten", NULL);
    EPD_4IN0E_TurnOnDisplay();
}

/******************************************************************************
function :  show 7 kind of color block
parameter:
******************************************************************************/
void EPD_4IN0E_Show7Block(void)
{
    unsigned long j, k;
    unsigned char const Color_seven[6] = 
    {EPD_4IN0E_BLACK, EPD_4IN0E_YELLOW, EPD_4IN0E_RED, EPD_4IN0E_BLUE, EPD_4IN0E_GREEN, EPD_4IN0E_WHITE};

    DEV_Debug_Log("show7Block:start", NULL);
    EPD_4IN0E_SendCommand(0x10);
    for(k = 0 ; k < 6; k ++) {
        for(j = 0 ; j < 20000; j ++) {
            EPD_4IN0E_SendData((Color_seven[k]<<4) |Color_seven[k]);
        }
    }
    DEV_Debug_Log("show7Block:dataWritten", NULL);
    EPD_4IN0E_TurnOnDisplay();
}

void EPD_4IN0E_Show(void)
{
    unsigned long k,o;
    unsigned char const Color_seven[6] = 
    {EPD_4IN0E_BLACK, EPD_4IN0E_YELLOW, EPD_4IN0E_RED, EPD_4IN0E_BLUE, EPD_4IN0E_GREEN, EPD_4IN0E_WHITE};

    UWORD Width, Height;
    Width = (EPD_4IN0E_WIDTH % 2 == 0)? (EPD_4IN0E_WIDTH / 2 ): (EPD_4IN0E_WIDTH / 2 + 1);
    Height = EPD_4IN0E_HEIGHT;
    k = 0;
    o = 0;

    DEV_Debug_Log("show:start", NULL);
    EPD_4IN0E_SendCommand(0x10);
    for (UWORD j = 0; j < Height; j++) {
        if((j > 10) && (j<50))
        for (UWORD i = 0; i < Width; i++) {
                EPD_4IN0E_SendData((Color_seven[0]<<4) |Color_seven[0]);
            }
        else if(o < Height/2)
        for (UWORD i = 0; i < Width; i++) {
                EPD_4IN0E_SendData((Color_seven[0]<<4) |Color_seven[0]);
            }
        
        else
        {
            for (UWORD i = 0; i < Width; i++) {
                EPD_4IN0E_SendData((Color_seven[k]<<4) |Color_seven[k]);
                
            }
            k++ ;
            if(k >= 6)
                k = 0;
        }
            
        o++ ;
        if(o >= Height)
            o = 0;
    }
    DEV_Debug_Log("show:dataWritten", NULL);
    EPD_4IN0E_TurnOnDisplay();
}

/******************************************************************************
function :  Sends the image buffer in RAM to e-Paper and displays
parameter:
******************************************************************************/
void EPD_4IN0E_Display(UBYTE *Image)
{
    UWORD Width, Height;
    Width = (EPD_4IN0E_WIDTH % 2 == 0)? (EPD_4IN0E_WIDTH / 2 ): (EPD_4IN0E_WIDTH / 2 + 1);
    Height = EPD_4IN0E_HEIGHT;

    DEV_Debug_Log("display:start", NULL);
    DEV_Debug_Preview(Image, (unsigned long)Width * Height);
    EPD_4IN0E_SendCommand(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_4IN0E_SendData(Image[i + j * Width]);
        }
    }
    DEV_Debug_Log("display:dataWritten", NULL);
    EPD_4IN0E_TurnOnDisplay();
}

/******************************************************************************
function :  Enter sleep mode
parameter:
******************************************************************************/
void EPD_4IN0E_Sleep(void)
{
    DEV_Debug_Log("sleep", NULL);
    EPD_4IN0E_SendCommand(0x07); // DEEP_SLEEP
    EPD_4IN0E_SendData(0XA5);
    // EPD_4IN0E_ReadBusyH();
}

